File: instrumentation/runner/src/main/java/org/junit/platform/runner/AndroidJUnitPlatformRunnerListener.java
package org.junit.platform.runner;

import android.annotation.SuppressLint;
import android.util.Log;

import org.junit.platform.engine.TestExecutionResult;
import org.junit.platform.launcher.TestExecutionListener;
import org.junit.platform.launcher.TestIdentifier;
import org.junit.runner.Description;
import org.junit.runner.notification.Failure;
import org.junit.runner.notification.RunNotifier;

import static de.mannodermaus.junit5.internal.ExtensionsKt.LOG_TAG;
import static org.junit.platform.engine.TestExecutionResult.Status.ABORTED;
import static org.junit.platform.engine.TestExecutionResult.Status.FAILED;

/**
 * Required, public extension to allow access to package-private RunnerListener class
 */
@SuppressLint("NewApi")
public final class AndroidJUnitPlatformRunnerListener implements TestExecutionListener {

  private final AndroidJUnitPlatformTestTree testTree;
  private final RunNotifier notifier;

  public AndroidJUnitPlatformRunnerListener(AndroidJUnitPlatformTestTree testTree, RunNotifier notifier) {
    this.testTree = testTree;
    this.notifier = notifier;
  }

  @Override
  public void executionStarted(TestIdentifier testIdentifier) {
    Description description = testTree.getDescription(testIdentifier);

    if (testIdentifier.isTest()) {
      notifier.fireTestStarted(description);
    }
  }

  @Override
  public void dynamicTestRegistered(TestIdentifier testIdentifier) {
    testTree.addDynamicDescription(testIdentifier, testIdentifier.getParentId().get());
  }

  @Override
  public void executionSkipped(TestIdentifier testIdentifier, String reason) {
    if (testIdentifier.isTest()) {
      fireTestIgnored(testIdentifier, reason);
    } else {
      testTree.getTestsInSubtree(testIdentifier).forEach(identifier -> fireTestIgnored(identifier, reason));
    }
  }

  private void fireTestIgnored(TestIdentifier testIdentifier, String reason) {
    Description description = testTree.getDescription(testIdentifier);
    this.notifier.fireTestIgnored(description);
    Log.w(LOG_TAG, testTree.getTestName(testIdentifier) + " is ignored. " + reason);
  }

  @Override
  public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
    Description description = testTree.getDescription(testIdentifier);
    TestExecutionResult.Status status = testExecutionResult.getStatus();
    if (status == ABORTED) {
      this.notifier.fireTestAssumptionFailed(toFailure(testExecutionResult, description));
    } else if (status == FAILED) {
      this.notifier.fireTestFailure(toFailure(testExecutionResult, description));
    }
    if (description.isTest()) {
      this.notifier.fireTestFinished(description);
    }
  }

  private Failure toFailure(TestExecutionResult testExecutionResult, Description description) {
    return new Failure(description, testExecutionResult.getThrowable().orElse(null));
  }
}

File: instrumentation/runner/src/main/kotlin/de/mannodermaus/junit5/internal/runners/AndroidJUnitPlatformRunnerListener.kt
package de.mannodermaus.junit5.internal.runners

import android.annotation.SuppressLint
import android.util.Log
import de.mannodermaus.junit5.internal.LOG_TAG
import org.junit.platform.engine.TestExecutionResult
import org.junit.platform.engine.reporting.ReportEntry
import org.junit.platform.launcher.TestExecutionListener
import org.junit.platform.launcher.TestIdentifier
import org.junit.platform.launcher.TestPlan
import org.junit.runner.Description
import org.junit.runner.notification.Failure
import org.junit.runner.notification.RunNotifier

/**
 * Required, public extension to allow access to package-private RunnerListener class
 */
@SuppressLint("NewApi")
internal class AndroidJUnitPlatformRunnerListener(
    private val testTree: AndroidJUnitPlatformTestTree,
    private val notifier: RunNotifier
) : TestExecutionListener {

    override fun reportingEntryPublished(testIdentifier: TestIdentifier?, entry: ReportEntry?) {
        // No-op, but must be declared to avoid AbstractMethodError
    }

    override fun testPlanExecutionStarted(testPlan: TestPlan) {
        notifier.fireTestSuiteStarted(testTree.suiteDescription)
    }

    override fun executionStarted(testIdentifier: TestIdentifier) {
        val description = testTree.getDescription(testIdentifier)
        if (testIdentifier.isTest) {
            notifier.fireTestStarted(description)
        }
    }

    override fun dynamicTestRegistered(testIdentifier: TestIdentifier) {
        testTree.addDynamicDescription(testIdentifier, testIdentifier.parentId.get())
    }

    override fun executionSkipped(testIdentifier: TestIdentifier, reason: String) {
        if (testIdentifier.isTest) {
            fireTestIgnored(testIdentifier, reason)
        } else {
            testTree.getTestsInSubtree(testIdentifier)
                .forEach { identifier ->
                    fireTestIgnored(identifier, reason)
                }
        }
    }

    private fun fireTestIgnored(testIdentifier: TestIdentifier, reason: String) {
        val description = testTree.getDescription(testIdentifier)
        notifier.fireTestIgnored(description)
        Log.w(LOG_TAG, testTree.getTestName(testIdentifier) + " is ignored. " + reason)
    }

    override fun executionFinished(
        testIdentifier: TestIdentifier,
        testExecutionResult: TestExecutionResult
    ) {
        val description = testTree.getDescription(testIdentifier)
        val status = testExecutionResult.status
        if (status == TestExecutionResult.Status.ABORTED) {
            notifier.fireTestAssumptionFailed(toFailure(testExecutionResult, description))
        } else if (status == TestExecutionResult.Status.FAILED) {
            notifier.fireTestFailure(toFailure(testExecutionResult, description))
        } else if (testIdentifier.isTest) {
            notifier.fireTestFinished(description)
        }
    }

    override fun testPlanExecutionFinished(testPlan: TestPlan) {
        notifier.fireTestSuiteFinished(testTree.suiteDescription)
    }

    private fun toFailure(
        testExecutionResult: TestExecutionResult,
        description: Description
    ): Failure = Failure(description, testExecutionResult.throwable.orElse(null))
}